### PR TITLE
Trie/stites

### PR DESCRIPTION
**Pull request by stites**
My first PR! The whole program is pretty darn slow to load. I'm guessing moving to Text would be a good idea, but my laptop is having some trouble installing it via cabal.

### CODE REVIEW

**Comment by int-index** [comment]
You should name the outer folder as the task and the inner folder after yourself, not vice versa.

**Review by vitcra — COMMENTED** [review]
This is very nice. The code clearly expresses what really happens in this exercise. I do not feel it is for me to approve this, since I am not yet at this level of Haskell.

**Comment by 444c43** [comment]
it's not compiling for me. i get the following error (in GHCI and trying to compile):

```
Main.hs:81:19:
    Not in scope: type constructor or class `Monoid'
    Perhaps you meant `Monad' (imported from Control.Monad)
```

do i need to install a specific library?

**Comment by stites** [comment]
Oops! I must have loaded Monoid in ghci and let that slip. @444c43 — Monoid is in the prelude at [`Data.Monoid`](https://hackage.haskell.org/package/base-4.9.0.0/docs/Data-Monoid.html).

**Comment by neongreen** [comment]
Please create branches in this repository instead of forking; this way it's easier for me to check them out.

File: week4/trie/stites/Main.hs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+------------------------------------------------------------------------
+-- Declare a Trie and useful typeclasses
+
+data Trie a
+  = Empty
+  | Node (Map a (Trie a))
+  deriving (Show)
+
+instance Ord a => Monoid (Trie a) where
+  mempty :: Trie a
```
Comment (neongreen) [inline week4/trie/stites/Main.hs:83]: Usually these signatures aren't written (and then you also don't need the `InstanceSigs` extension).

File: week4/trie/stites/Main.hs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+main :: IO ()
+main = do
+  contents <- readFile filepath
+  let t = fromFileContents contents :: Trie Char
+  putStrLn $ "Trie created. There are " ++ show (size t)  ++ " nodes."
+  forever $ do
+    putStr "> "
+    query <- getLine
+    putStrLn . renderQueryResults .  queryTrie t $ query
+    return ()
```
Comment (neongreen) [inline week4/trie/stites/Main.hs:60]: `return ()` isn't needed here (I guess it's a leftover and at some previous point it was needed though)

File: week4/trie/stites/Main.hs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+import qualified Data.Map.Strict as M
+
+filepath :: FilePath
+filepath = "data/words"
+
+main :: IO ()
+main = do
+  contents <- readFile filepath
+  let t = fromFileContents contents :: Trie Char
+  putStrLn $ "Trie created. There are " ++ show (size t)  ++ " nodes."
```
Comment (neongreen) [inline week4/trie/stites/Main.hs:55]: “Trie created” will be printed before the trie is actually created, thanks to laziness. To only print it afterwards, do something like this:

```
let t = fromFileContents contents :: Trie Char
    len = size t
len `seq` putStrLn ("Trie created. There are " ++ show len ++ " nodes.")
```

File: week4/trie/stites/Main.hs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+------------------------------------------------------------------------
+-- Declare a Trie and useful typeclasses
+
+data Trie a
+  = Empty
+  | Node (Map a (Trie a))
+  deriving (Show)
+
+instance Ord a => Monoid (Trie a) where
```
Comment (neongreen) [inline week4/trie/stites/Main.hs:82]: Nice job making a `Monoid` instance. Usually people forget to create instances even when it's possible to do so.

File: week4/trie/stites/Main.hs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+-- A Trie can be constructed from a list
+fromList :: Ord a => [a] -> Trie a
+fromList    []  = Empty
+fromList (a:as) = Node . M.singleton a . fromList $ as
+
+-- A Trie can be deconstructed to a list of lists
+toList :: Trie a -> [[a]]
+toList Empty = [[]]
+toList (Node m) = concat . M.elems . M.mapWithKey go $ m
```
Comment (neongreen) [inline week4/trie/stites/Main.hs:99]: You can just write `concat . M.mapWithKey go`. After FTP (the Foldable/Traversable Proposal) was implemented, `concat` works on much more things.

**Review by neongreen — APPROVED** [review]

**Pull request merged**